pkg/clients/container/cluster_test.go

```diff
@@ -0,0 +1,1593 @@
+			if diff := cmp.Diff(tc.want.WorkloadIdentityConfig, tc.args.cluster.WorkloadIdentityConfig); diff != "" {
+				t.Errorf("GenerateWorkloadIdentityConfig(...): -want, +got:\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestLateInitializeSpec(t *testing.T) {
+	type args struct {
+		cluster *container.Cluster
+		params  *v1beta1.GKEClusterParameters
+	}
+	type want struct {
+		params *v1beta1.GKEClusterParameters
+	}
+	tests := map[string]struct {
+		args args
+		want want
+	}{
+		"SomeFilled": {
+			args: args{
+				cluster: cluster(func(c *container.Cluster) {
+					c.AddonsConfig = &container.AddonsConfig{
+						HttpLoadBalancing: &container.HttpLoadBalancing{
+							Disabled: true,
+						},
+					}
+					c.IpAllocationPolicy = &container.IPAllocationPolicy{
+						ClusterIpv4CidrBlock: "0.0.0.0/0",
+					}
+				}),
+				params: params(),
+			},
+			want: want{
+				params: params(func(p *v1beta1.GKEClusterParameters) {
+					p.AddonsConfig = &v1beta1.AddonsConfig{
+						HTTPLoadBalancing: &v1beta1.HTTPLoadBalancing{
+							Disabled: true,
+						},
+					}
+					p.IPAllocationPolicy = &v1beta1.IPAllocationPolicy{
+						ClusterIpv4CidrBlock: gcp.StringPtr("0.0.0.0/0"),
+					}
+				}),
+			},
+		},
+		"AllFilled": {
+			args: args{
+				cluster: cluster(),
+				params:  params(),
+			},
+			want: want{
+				params: params(),
+			},
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			LateInitializeSpec(tc.args.params, *tc.args.cluster)
+			if diff := cmp.Diff(tc.want.params, tc.args.params); diff != "" {
+				t.Errorf("LateInitializeSpec(...): -want, +got:\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestIsUpToDate(t *testing.T) {
+	type args struct {
+		cluster *container.Cluster
+		params  *v1beta1.GKEClusterParameters
+	}
+	type want struct {
+		upToDate bool
+	}
+	tests := map[string]struct {
+		args args
+		want want
+	}{
+		"UpToDate": {
+			args: args{
+				cluster: cluster(),
+				params:  params(),
+			},
+			want: want{
+				upToDate: true,
+			},
+		},
+		"UpToDateIgnoreRefs": {
+			args: args{
+				cluster: cluster(),
+				params: params(func(p *v1beta1.GKEClusterParameters) {
+					p.NetworkRef = &v1beta1.NetworkURIReferencerForGKECluster{
+						NetworkURIReferencer: v1alpha3.NetworkURIReferencer{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: "my-network",
+							},
+						},
+					}
+				}),
+			},
+			want: want{
+				upToDate: true,
+			},
+		},
+		"NeedsUpdate": {
+			args: args{
+				cluster: cluster(func(c *container.Cluster) {
+					c.AddonsConfig = &container.AddonsConfig{
+						HttpLoadBalancing: &container.HttpLoadBalancing{
+							Disabled: true,
+						},
+					}
+					c.IpAllocationPolicy = &container.IPAllocationPolicy{
+						ClusterIpv4CidrBlock: "0.0.0.0/0",
+					}
+				}),
+				params: params(),
+			},
+			want: want{
+				upToDate: false,
+			},
+		},
+		"NoUpdateNotBootstrapNodePool": {
```

Comment:
This and the subsequent test demonstrate how the update controller handles deletion of the default "bootstrap" node pool.